conda_lock/src_parser/aggregation.py

```diff
@@ -56,7 +65,12 @@ def aggregate_lock_specs(
     )
 
 
-def unify_package_sources(collections: List[List[Channel]]) -> List[Channel]:
+PackageSource = TypeVar("PackageSource", Channel, PipRepository)
+
+
+def unify_package_sources(
+    collections: List[List[PackageSource]],
+) -> List[PackageSource]:
```

Comment:
This is really slick if I'm following correctly, because I would have done `Union[List[List[Channel]], List[List[PipRepository]]]` and then used `@overload` to ensure the right return type; much messier!

As a `TypeVar` this prevents mixed lists, right (e.g. `List[List[Union[Channel, PipRepository]]]`)? (Sorry, I'm unable to install `mypy` to check myself.)

Reply:
Finally managed to install `mypy` and confirmed it. Thanks for the awesome trick!

Reply:
Yes that's exactly correct. The type variable of the argument defines the type argument of the return value. So:

```python
repos: list[list[PipRepository]]
unify_package_sources(repos)  # return type inferred as list[PipRepository]
```